include/cuco/detail/hyperloglog/hyperloglog_ref.cuh

```diff
@@ -0,0 +1,189 @@
+        this->storage_[i] = max(this->storage_[i], other.storage_[i]);
+      } else if constexpr (Scope == cuda::thread_scope_block) {
+        atomicMax_block(this->storage_.data() + i, other.storage_[i]);
+      } else if constexpr (Scope == cuda::thread_scope_device) {
+        atomicMax(this->storage_.data() + i, other.storage_[i]);
+      } else if constexpr (Scope == cuda::thread_scope_system) {
+        atomicMax_system(this->storage_.data() + i, other.storage_[i]);
+      } else {
+        static_assert(cuco::dependent_false<decltype(Scope)>, "Unsupported thread scope");
+      }
+    }
+
+    // TODO remove test code
+    /*
+    auto vec4 = reinterpret_cast<int4 const*>(other.storage_.data());
+    // #pragma unroll 2
+    for (int i = group.thread_rank(); i < (this->storage_.size() / 4); i += group.size()) {
+      auto const items = vec4[i];
+      if constexpr (Scope == cuda::thread_scope_thread) {
+        auto max_vec4  = reinterpret_cast<int4*>(this->storage_.data());
+        auto max_items = max_vec4[i];
+        max_items.x    = max(max_items.x, items.x);
+        max_items.y    = max(max_items.y, items.y);
+        max_items.z    = max(max_items.z, items.z);
+        max_items.w    = max(max_items.w, items.w);
+        max_vec4[i]    = max_items;
+      } else if constexpr (Scope == cuda::thread_scope_block) {
+        atomicMax_block(this->storage_.data() + (i * 4 + 0), items.x);
+        atomicMax_block(this->storage_.data() + (i * 4 + 1), items.y);
+        atomicMax_block(this->storage_.data() + (i * 4 + 2), items.z);
+        atomicMax_block(this->storage_.data() + (i * 4 + 3), items.w);
+      } else if constexpr (Scope == cuda::thread_scope_device) {
+        atomicMax(this->storage_.data() + (i * 4 + 0), items.x);
+        atomicMax(this->storage_.data() + (i * 4 + 1), items.y);
+        atomicMax(this->storage_.data() + (i * 4 + 2), items.z);
+        atomicMax(this->storage_.data() + (i * 4 + 3), items.w);
+      } else if constexpr (Scope == cuda::thread_scope_system) {
+        atomicMax_system(this->storage_.data() + (i * 4 + 0), items.x);
+        atomicMax_system(this->storage_.data() + (i * 4 + 1), items.y);
+        atomicMax_system(this->storage_.data() + (i * 4 + 2), items.z);
+        atomicMax_system(this->storage_.data() + (i * 4 + 3), items.w);
+      } else {
+        static_assert(cuco::dependent_false<decltype(Scope)>, "Unsupported thread scope");
+      }
+    }
+    */
+  }
+
+  [[nodiscard]] __device__ std::size_t estimate(
+    cooperative_groups::thread_block const& group) const noexcept
+  {
+    __shared__ cuda::atomic<fp_type, cuda::thread_scope_block> block_sum;
+    __shared__ cuda::atomic<int, cuda::thread_scope_block> block_zeroes;
+    __shared__ std::size_t estimate;
+
+    // TODO is this needed?
+    if (group.thread_rank() == 0) {
+      block_sum.store(0, cuda::std::memory_order_relaxed);
+      block_zeroes.store(0, cuda::std::memory_order_relaxed);
+    }
+    group.sync();
```

Comment:
suggestion: This is UB as it is using member functions from an uninitialized object. It should use placement-new or `cuda::std::construct_at` to construct the object.

Reply:
Good catch! I'm unable to locate `cuda::std::construct_at`. Shouldn't that exist in something like `<cuda/std/memory>`? I can of course stick with placement new for now.

Reply:
@miscco don't we have `construct_at` publicly exposed?

Reply:
`construct_at` is only available from C++20 onwards. You could spell it `cuda::std::__construct_at` and that would always work.

The issue is that `construct_at` has a special blessed meaning in the standard, which is why I did not enable it in earlier modes. We can happily backport it though

Reply:
I went with placement-new for now. If we decide on backporting `cuda::std::construct_at` to C++17 I will happily refactor this line :)